### PR TITLE
[ENH] Multioutput capability for all time series classifiers and regressors, broadcasting and tag

### PR DESCRIPTION
**Pull request by Vasudeva-bit**
#### Reference Issues/PRs

Related to #5182, #5286

Depends on https://github.com/sktime/sktime/pull/5651, which should be merged first.

#### What does this implement/fix? Explain your changes.

Adds broadcasting support to let classifiers and regressors handle `multioutput` data.

#### What should a reviewer concentrate their feedback on?

Kindly give feedback on whether I am going in right direction.

##### For all contributions
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

File: sktime/classification/base.py [hunk 1]
```diff
@@ -131,7 +134,213 @@ def __rmul__(self, other):
...
+            self._fit_instance(X, y, metadata)
+            self.estimators = [self]
+            return self
+        else:
+            vdf = VectorizedDF(
+                X=[X] * len(y.columns),
+                y=[(lambda col: y[col])(col) for col in y.columns],
+            )
+            self.estimators = vdf.vectorize_est(estimator=self, method="clone")
+            for col, est in zip(y.columns, self.estimators):
```
Comment (fkiraly) [inline sktime/classification/base.py:171]: should this not be replaced by `vectorize_est`, `method="fit"`?

Comment (Vasudeva-bit) [reply sktime/classification/base.py:171]: I thought of two ways of approaching this issue:
1. Create instances of unfitted estimator using `method='clone'`, fit each of them individually, predict individually.
2. Use `vectorize_est` with `method='fit'` and `method='predict'` across the columns of y.

I am not sure which one is better?

Comment (fkiraly) [reply sktime/classification/base.py:171]: 2. is the programmatic way to do 1. in sktime, the individual fits/predicts are stored in the returned data frame, and `vectorize_est` also gives access to arbitrary parallel backends

**Comment by Vasudeva-bit** [comment]
It has been a while since I last worked on this issue, I would like to get a quick clarification before starting again.
Kindly help me know which of the below methods is better:
1. As discussed [here](https://github.com/sktime/sktime/issues/5182) i.e., extend required classifiers and regressors for multioutput `y`.

> I suppose setting false as a default has only be done in the base class.
> 
> Only classes need to be adapted that support multitarget prediction natively.

2. Otherwise, modify only classification and regression base classes similar to `sktime/transformations/base.py` i.e., base class handles multioutput based on the tag `"capability:multioutput"` assigned in estimator (classifier/regressor) classes.
@fkiraly

**Comment by fkiraly** [comment]
Yes - I would recommend to start with no.2 to get the framework ready, and only then add multioutput support for some classifiers, e.g., the neural networks. That way, no.1 will also be a test case for the framework extensions in no.2.

**Comment by Vasudeva-bit** [comment]
Kindly review the code and help me with the following:
1. If the format looks good, I shall extend regression base class as well.
2. I need to set the tag `"capability:multioutput"` to `True ` to all the estimators that support it. Kindly mention estimators that you know can support multioutput. I will try to figure the rest. Now, only `CNNClassifier`'s tag is set to `True`.
> add multioutput support for some classifiers, e.g., the neural networks
    
Once I am done with the above two, I shall continue with this. @fkiraly

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!
This looks like it works almost.

There's something upset with the tags, can you check the tests?

Also, non-blocking, do you think we can avoid introducing an "instance" copy of every method, without making the code less DRY? I don't see it, but it might be worth thinking briefly about this.

**Comment by Vasudeva-bit** [comment]
> There's something upset with the tags, can you check the tests?

> Also, non-blocking, do you think we can avoid introducing an "instance" copy of every method, without making the code less DRY? I don't see it, but it might be worth thinking briefly about this.

I shall look into these.

**Comment by Vasudeva-bit** [comment]
I have made changes, kindly review them.
Btw why are tests not triggering?

**Comment by fkiraly** [comment]
> Btw why are tests not triggering?

There are some merge conflicts you need to resolve first, for this you need to merge `main` into your branch and resolve the conflicts with a diff editor or similar.

File: sktime/classification/tests/test_base.py [hunk 2]
```diff
@@ -154,15 +152,6 @@ def test_base_classifier_fit():
     with pytest.raises(ValueError, match=multivariate_message):
         result = dummy.fit(test_X4, test_y1)
     assert result is dummy
-    # Raise a specific error if y is in a 2D matrix (1,cases)
```
[the remark this answers is not in the document]
Comment (Vasudeva-bit) [reply sktime/classification/tests/test_base.py:157]: Will add that.

File: sktime/classification/base.py [hunk 3]
```diff
@@ -320,14 +353,124 @@ def fit_predict(self, X, y, cv=None, change_state=True) -> np.ndarray:
...
+        else:
+            if self._classifiers_ is not None:
+                classifiers_ = self._classifiers_
+            y_pred = self._y_vec.vectorize_est(
+                classifiers_,
+                method=methodname,
+                # return_type="list",
+                X=X,
+                args={"y": y} if y is not None else {},
+                *kwargs,  # contains X inside
```
Comment (fkiraly) [inline sktime/classification/base.py:443]: use `**kwargs`

File: sktime/classification/base.py [hunk 4]
```diff
@@ -755,12 +855,6 @@ def _check_classifier_input(
                     f"Mismatch in number of cases. Number in X = {n_cases} nos in y = "
                     f"{n_labels}"
                 )
-            if isinstance(y, np.ndarray):
-                if y.ndim > 1:
```
[the remark this answers is not in the document]
Comment (Vasudeva-bit) [reply sktime/classification/base.py:759]: Yes, I get it.

File: sktime/classification/base.py (hunk 3 above)
Comment (Vasudeva-bit) [reply sktime/classification/base.py:443]: I overlooked that.

**Comment by Vasudeva-bit** [comment]
I did broadcast `BaseRegressor`. Now, I need to give proper value to tag `capability:multioutput` to each classifier and regressor. I believe some of the estimators (especially neural network based) are capable of dealing with multioutput. Kindly let me know if any other estimator is capable of dealing with multioutput. I also noticed that there isn't any `test_base.py` to test `BaseRegressor`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
The error seems to be that you introduced a character somewhere that is problematic.

Common culprits are wrong versions of the double quotes or similar. I cannot immediately spot anything, but maybe your linting could find it?

File: sktime/regression/tests/test_all_regressors.py [hunk 5]
```diff
@@ -71,12 +71,11 @@ def test_regressor_output(self, estimator_instance, scenario):
         # we use check_is_scitype to get the number instances in X_new
         #   this is more robust against different scitypes in X_new
         _, _, X_new_metadata = check_is_scitype(X_new, "Panel", return_metadata=True)
-        X_new_instances = X_new_metadata["n_instances"]
```
Comment (fkiraly) [inline sktime/regression/tests/test_all_regressors.py:74]: why are we changing this?

File: sktime/regression/base.py [hunk 6]
```diff
@@ -151,32 +157,43 @@ def fit(self, X, y):
...
+                    )
+
+            self._fit(X, y)
+            self.fit_time_ = int(round(time.time() * 1000)) - start
+        else:
+            # otherwise we call the vectorized version of fit
+            self._vectorize("fit", X=X, y=y)
 
-        self._fit(X, y)
-        self.fit_time_ = int(round(time.time() * 1000)) - start
```
Comment (fkiraly) [inline sktime/regression/base.py:179]: I would keep the `fit_time_` outside the loop.

Comment (Vasudeva-bit) [reply sktime/regression/base.py:179]: That makes sense. I think `start` should also be outside.

File: sktime/regression/tests/test_all_regressors.py (hunk 5 above)
Comment (Vasudeva-bit) [reply sktime/regression/tests/test_all_regressors.py:74]: `X_meta_instances` no longer needed, I modified another line below. I shall add it back if it's required.

Comment (fkiraly) [reply sktime/regression/tests/test_all_regressors.py:74]: I meant, we still should check for the right length here, i.e., leave it in - I do not see how the expected output format would change in the scenario.

Comment (Vasudeva-bit) [reply sktime/regression/tests/test_all_regressors.py:74]: Yeah, I see it. Now the shape check should be modified according to dim of y i.e., one case is when dim is 1, and another when dim is 2.

File: sktime/classification/tests/test_all_classifiers.py [hunk 7]
```diff
@@ -86,13 +86,20 @@ def test_classifier_output(self, estimator_instance, scenario):
 
         # check predict
         assert isinstance(y_pred, np.ndarray)
-        assert y_pred.shape == (X_new_instances,)
+        assert y_pred.shape == (X_new_instances,) or y_pred.shape == (
+            X_new_instances,
+            y_pred.ndim,
```
Comment (fkiraly) [inline sktime/classification/tests/test_all_classifiers.py:91]: this does not look right, `ndim` is 1 or 2 always.

File: sktime/classification/tests/test_all_classifiers.py [hunk 8]
```diff
@@ -86,13 +86,20 @@ def test_classifier_output(self, estimator_instance, scenario):
...
+        )
         assert np.all(np.isin(np.unique(y_pred), np.unique(y_train)))
 
         # check predict proba (all classifiers have predict_proba by default)
         y_proba = scenario.run(estimator_instance, method_sequence=["predict_proba"])
         assert isinstance(y_proba, np.ndarray)
-        assert y_proba.shape == (X_new_instances, n_classes)
+        assert y_proba.shape == (X_new_instances, n_classes) or y_proba.shape == (
+            X_new_instances,
```
Comment (fkiraly) [inline sktime/classification/tests/test_all_classifiers.py:99]: is this the sklearn convention?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

Small comments:
* you use `y_pred.ndim` for number of variables, in tests, but that is not correct - that is the number of dimensions, not variables.
* I would suggest to add a multivariate test case in the tests where additional checks happen.

File: sktime/regression/tests/test_all_regressors.py [hunk 9]
```diff
@@ -78,5 +78,8 @@ def test_regressor_output(self, estimator_instance, scenario):
 
         # check predict
         assert isinstance(y_pred, np.ndarray)
-        assert y_pred.shape == (X_new_instances,)
+        assert y_pred.shape == (X_new_instances,) or y_pred.shape == (
+            X_new_instances,
+            y_pred.ndim,
```
Comment (fkiraly) [inline sktime/regression/tests/test_all_regressors.py:83]: same as above

File: sktime/classification/tests/test_all_classifiers.py (hunk 7 above)
Comment (Vasudeva-bit) [reply sktime/classification/tests/test_all_classifiers.py:91]: That has to be `y_train.shape[1]` iff `y_train.ndim > 1`. I hope I got it.

File: sktime/classification/tests/test_all_classifiers.py (hunk 8 above)
Comment (Vasudeva-bit) [reply sktime/classification/tests/test_all_classifiers.py:99]: I am not sure about that. I just modified the check based on the previous one, except that now it also checks for multioutput.

**Comment by Vasudeva-bit** [comment]
> * I would suggest to add a multivariate test case in the tests where additional checks happen.

Do you mean to add multivariate test case separately to classifier and regressor classes. Also kindly help me understand what additional checks are. Are they dummy I/O tests, no side effects, dimension/shape checks etc.?

**Comment by fkiraly** [comment]
> Are they dummy I/O tests, no side effects, dimension/shape checks etc.?

Yes, I was thinking only of input/output, with multivariate `y`, and possibly check the `classifiers_` attribute, etc.

**Comment by fkiraly** [comment]
@Vasudeva-bit, I've taken over this PR to ensure this gets over the finish line for the release.

Changes I intend:

* programmatic input checks and conversions for `y` using the `Table` scitype, simplify multiple input checks
* minimize change footprint in `_fit`, `_predict` etc methods for the non-vectorized case

**Comment by Vasudeva-bit** [comment]
Extremely sorry for the delay, I was busy with other commitments last week. Kindly let me know if I could do anything. I am available on 23, 24, 25 completely. Thanks.

**Comment by fkiraly** [comment]
>Kindly let me know if I could do anything.

Review would be appreciated, e.g., is this in line with your intentions/needs, and is it correct.

File: sktime/classification/tests/test_all_classifiers.py (hunk 7 above)
Comment (Vasudeva-bit) [reply sktime/classification/tests/test_all_classifiers.py:91]: This change is yet to be done.

File: sktime/regression/tests/test_all_regressors.py (hunk 9 above)
Comment (Vasudeva-bit) [reply sktime/regression/tests/test_all_regressors.py:83]: `y_pred.ndim` should be `y_train.shape[1]` iff `y_train.ndim > 1`.

**Comment by Vasudeva-bit** [comment]
is `test_base.py` required for regression base class?

File: sktime/classification/base.py
Comment (Vasudeva-bit) [inline sktime/classification/base.py:1]: One of the ensemble classifiers `_weighted.py` has `self.classifiers_` attribute which might conflict with `self.classifiers_` in classification's `base.py`. [Here](https://github.com/sktime/sktime/blob/734d63a2ef21009c0d35c03d59d504e7af4cff8b/sktime/classification/ensemble/_weighted.py#L127).

File: sktime/regression/base.py
Comment (Vasudeva-bit) [inline sktime/regression/base.py:1]: Inside `_vectorize` method, variables `self._classifiers_`, `self.classifiers_` have to be changed to `self._regressors_`, `self.regressors_`. It was my mistake.

File: sktime/classification/base.py
Comment (fkiraly) [reply sktime/classification/base.py:1]: Yes, I noticed - it seems to be resolved though if `__init__` does not set this.
I was trying to be consistent with forecasting, which has `forecasters_`.

**Comment by fkiraly** [comment]
> for estimators that are capable of multi-output needs to be set to `True`.

I do not know any estimators that currently are genuinely capable of multioutput. Or, am I mistaken?

**Comment by Vasudeva-bit** [comment]
> I do not know any estimators that currently are genuinely capable of multioutput. Or, am I mistaken?

I think a few neural network based estimators like `cnn.py` can handle multioutput. I am not sure, kindly refer #

**Comment by fkiraly** [comment]
> I think a few neural network based estimators like cnn.py can handle multioutput

in-principle certainly, but, right at the moment? Do these not break, if they get multivariate `y`?

**Comment by Vasudeva-bit** [comment]
> in-principle certainly, but, right at the moment? Do these not break, if they get multivariate `y`?

I think they break. I will check it again.

**Comment by fkiraly** [comment]
> I think they break. I will check that again.

Well, if they break, they should not get the `multioutput` tag until they are upgraded.

File: sktime/regression/base.py
Comment (fkiraly) [reply sktime/regression/base.py:1]: yes, will get to that

**Review by fkiraly — DISMISSED** [review]
All updated now. I removed the tags from classifiers, we can do that separately.
